linux/ibus-keyman/src/engine.c

```diff
@@ -558,160 +560,113 @@ static void commit_string(IBusKeymanEngine *keyman, const gchar *string)
+    g_message("%s: compliant app: deleting surrounding text %d codepoints", __FUNCTION__, code_points_to_delete);
+    ibus_engine_delete_surrounding_text(engine, -code_points_to_delete, code_points_to_delete);
   } else {
-    g_message(
-        "DAR: %s - client_capabilities=%x, %x", __FUNCTION__, engine->client_capabilities, IBUS_CAP_SURROUNDING_TEXT);
-
-    if (client_supports_surrounding_text(engine)) {
-      keyman->consecutive_backspaces++;
-      g_message("%s: compliant app: increment consecutive backspaces to %d", __FUNCTION__, keyman->consecutive_backspaces);
-    } else {
-      g_message("%s: non-compliant app: forwarding backspace", __FUNCTION__);
-      forward_backspace(keyman, 0);
+    g_message("%s: non-compliant app: forwarding %d backspaces", __FUNCTION__, code_points_to_delete);
+    while (code_points_to_delete > 0) {
+      ibus_engine_forward_key_event(engine, KEYMAN_BACKSPACE_KEYSYM, KEYMAN_BACKSPACE, 0);
+      code_points_to_delete--;
     }
   }
-  return TRUE;
-}
-
-static gboolean
-finish_backspace_action(
-  IBusEngine *engine
-) {
-  IBusKeymanEngine *keyman = (IBusKeymanEngine *)engine;
-  if (keyman->consecutive_backspaces <= 0)
-    return TRUE;
-
-  g_assert(client_supports_surrounding_text(engine));
-  g_message("%s: deleting surrounding text %d char", __FUNCTION__, keyman->consecutive_backspaces);
-  ibus_engine_delete_surrounding_text(engine, -keyman->consecutive_backspaces, keyman->consecutive_backspaces);
-  keyman->consecutive_backspaces = 0;
-  return TRUE;
 }
 
-static gboolean
-process_persist_action(
-  IBusKeymanEngine *keyman,
-  const km_core_action_item *action_item
-) {
-  // Save keyboard option
-  if (!action_item->option)
-    return TRUE;
-
-  // Allocate for 1 option plus 1 pad struct of 0's
-  km_core_option_item *keyboard_opts = g_new0(km_core_option_item, 2);
-  memmove(&(keyboard_opts[0]), action_item->option, sizeof(km_core_option_item));
-  km_core_status event_status = km_core_state_options_update(keyman->state, keyboard_opts);
-  if (event_status != KM_CORE_STATUS_OK) {
-    g_warning("%s: problem saving option for km_core_keyboard", __FUNCTION__);
+static void
+process_persist_action(IBusEngine *engine, km_core_option_item *persist_options) {
+  if (persist_options == NULL) {
+    return;
   }
-  g_free(keyboard_opts);
 
-  // Put the keyboard option into DConf
-  if (action_item->option->key != NULL && action_item->option->value != NULL) {
-    g_message("%s: Saving keyboard option to DConf", __FUNCTION__);
-    // Load the current keyboard options from DConf
-    keyman_put_options_todconf(
-        keyman->kb_name, keyman->kb_name, (gchar *)action_item->option->key, (gchar *)action_item->option->value);
+  IBusKeymanEngine *keyman = (IBusKeymanEngine *)engine;
+  for (km_core_option_item *option = persist_options; !is_core_options_end(option); option++) {
+    // Put the keyboard option into DConf
+    if (option->key != NULL && option->value != NULL) {
```

Comment:
This should probably be a g_assert as well; we should never be passing invalid data to you from Core.